Design/Algorithm_ease_of_use.md

```diff
@@ -0,0 +1,88 @@
+Motivation
+==========
+
+Currently code (especially workflow algorithms) will be full of things
+like
+
+```C
+double tolerance = getProperty("CompressTolerance");
+m_outputEW = boost::dynamic_pointer_cast<EventWorkspace>(m_outputW);
+if ((m_outputEW) && (tolerance > 0.)) {
+  g_log.information() << "running CompressEvents(Tolerance=" << tolerance
+                      << ")\n";
+  API::IAlgorithm_sptr compressAlg = createChildAlgorithm("CompressEvents");
+  compressAlg->setProperty("InputWorkspace", m_outputEW);
```

Comment:
I have a problem with member variables `m_` i.e. `m_outputEW` being used from any dynamic property. exec should be stateless in the context of the Algorithm object unless it's absolutely necessary to cache things for performance reasons. Can we consider fixing this somehow, or at least making it harder to do. How about `exec` becoming a free function called by the algorithm, where we pass in all the Properties needed?